pkg/virt-controller/watch/vm.go

```diff
@@ -1648,31 +1648,79 @@ func (c *VMController) enqueueVm(obj interface{}) {
 	c.Queue.Add(key)
 }
 
+func (c *VMController) getPatchFinalizerOps(oldObj v1.Object, newObj v1.Object) ([]string, error) {
+	var ops []string
+	newFinalizers, err := json.Marshal(newObj.GetFinalizers())
+	if err != nil {
+		return ops, err
+	}
+
+	oldFinalizers, err := json.Marshal(oldObj.GetFinalizers())
+	if err != nil {
+		return ops, err
+	}
+
+	ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/metadata/finalizers", "value": %s }`, string(oldFinalizers)))
+	ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/finalizers", "value": %s }`, string(newFinalizers)))
+	return ops, nil
+}
+
 func (c *VMController) removeVMIFinalizer(vmi *virtv1.VirtualMachineInstance) error {
-	vmiCopy := vmi.DeepCopy()
-	controller.RemoveFinalizer(vmiCopy, virtv1.VirtualMachineControllerFinalizer)
+	newVmi := vmi.DeepCopy()
+	controller.RemoveFinalizer(newVmi, virtv1.VirtualMachineControllerFinalizer)
 
-	if equality.Semantic.DeepEqual(vmi.Finalizers, vmiCopy.Finalizers) {
+	if equality.Semantic.DeepEqual(vmi.Finalizers, newVmi.Finalizers) {
 		return nil
 	}
 
 	log.Log.V(3).Object(vmi).Infof("VMI is in a final state. Removing VM controller finalizer")
 
-	ops := []string{}
-	newFinalizers, err := json.Marshal(vmiCopy.Finalizers)
-	if err != nil {
-		return err
-	}
-	oldFinalizers, err := json.Marshal(vmi.Finalizers)
+	ops, err := c.getPatchFinalizerOps(vmi, newVmi)
 	if err != nil {
 		return err
 	}
-	ops = append(ops, fmt.Sprintf(`{ "op": "test", "path": "/metadata/finalizers", "value": %s }`, string(oldFinalizers)))
-	ops = append(ops, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/finalizers", "value": %s }`, string(newFinalizers)))
+
 	_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, controller.GeneratePatchBytes(ops), &v1.PatchOptions{})
 	return err
 }
 
+func (c *VMController) removeVMFinalizer(vm *virtv1.VirtualMachine, finalizer string) (*virtv1.VirtualMachine, error) {
+	newVm := vm.DeepCopy()
+	controller.RemoveFinalizer(newVm, finalizer)
+
+	if equality.Semantic.DeepEqual(vm.Finalizers, newVm.Finalizers) {
+		return vm, nil
+	}
+
+	log.Log.V(3).Object(vm).Infof("VM is in a final state. Removing VM controller finalizer")
+
+	ops, err := c.getPatchFinalizerOps(vm, newVm)
+	if err != nil {
+		return vm, err
+	}
+
+	vm, err = c.clientset.VirtualMachine(vm.Namespace).Patch(vm.Name, types.JSONPatchType, controller.GeneratePatchBytes(ops), &v1.PatchOptions{})
+	return vm, err
+}
+
+func (c *VMController) addVMFinalizer(vm *virtv1.VirtualMachine, finalizer string) (*virtv1.VirtualMachine, error) {
+	newVm := vm.DeepCopy()
```

Comment:
I would like us to avoid unneeded `DeepCopy` at least in the hot path. Let's just iterate over the `Finalizers`.

Reply:
Nice catch!

A bit off-topic, but now that golang has generics I would have expected that general purpose copy/equality functions would be introduced, if not in the standard library then at least by some external one. Is there anything like that out there that we can use in Kubevirt?